qemu/tests/migration_with_vsock.py

```diff
@@ -0,0 +1,114 @@
+    """
+    Vsock migration test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    S1):
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+    7. Reboot guest with differnt guest CID
+    8. Migration
+    9. Check guest session exited and close host session
+    10. repeat step 3 to 6
+    11. Send data from guest, nc-vsock -l $port < tmp_file
+    12. Receive data from host, nc-vsock $guest_cid $port
+    13. Do ping-pong migration 3 times
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def ping_pong_migration(repeat_times):
+        """ Do ping pong migration. """
+        mig_timeout = float(params.get("mig_timeout", "3600"))
+        mig_protocol = params.get("migration_protocol", "tcp")
+        mig_cancel_delay = int(params.get("mig_cancel") == "yes") * 2
+        inner_funcs = ast.literal_eval(params.get("migrate_inner_funcs", "[]"))
+        capabilities = ast.literal_eval(params.get("migrate_capabilities", "{}"))
+        for i in range(repeat_times):
+            if i % 2 == 0:
+                logging.info("Round %s ping..." % str(i / 2))
+            else:
+                logging.info("Round %s pong..." % str(i / 2))
+            vm.migrate(
+                mig_timeout,
+                mig_protocol,
+                mig_cancel_delay,
+                migrate_capabilities=capabilities,
+                mig_inner_funcs=inner_funcs,
+                env=env,
+            )
+
+    def input_character_vsock():
+        connected_str = r"Connection from cid*"
+        nc_vsock_listen(nc_vsock_bin, port, session)
+        host_vsock_session = nc_vsock_connect(nc_vsock_bin, guest_cid, port)
+        send_data = "Hello world"
+        check_received_data(test, session, connected_str)
+        error_context.context('Input "Hello world" to vsock.', logging.info)
+        host_vsock_session.sendline(send_data)
+        check_received_data(test, session, send_data)
+        return host_vsock_session
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    nc_vsock_bin = compile_nc_vsock(test, vm, session)
+    vsock_dev = params["vsocks"].split()[0]
+    guest_cid = vm.devices.get(vsock_dev).get_param("guest-cid")
+    port = random.randrange(1, 6000)
+    host_vsock_session = input_character_vsock()
+    guest_cid = int(vm.devices.get(vsock_dev).get_param("guest-cid")) + 1
+    vm.devices.get(vsock_dev).set_param("guest-cid", guest_cid)
+    session.close()
+    # reboot with different guest-cid
+    vm.reboot()
+    # do migration
+    ping_pong_migration(1)
+    session = vm.wait_for_login()
+    if not session.cmd_status("ss --vsock sport eq %s " % port):
```

Comment:
@michiboo  Unfortunately,  no matter get the vsock socket with the given *port* or not, the return value always is `0`.
This does not work.
```
# ss --vsock sport eq 3785
Netid       State       Recv-Q        Send-Q               Local Address:Port               Peer Address:Port             
v_str       ESTAB       0             0                                3:3785                          2:2825511742       
# echo $?
0
# ss --vsock sport eq 3785
Netid        State        Recv-Q        Send-Q                Local Address:Port                 Peer Address:Port        
# echo $?
0
```

Reply:
I think remove NOT in if statement can not fix the problem. The return value of ss cmd is always 0, no matter find the port or not.

Reply:
Hi I changed to session.cmd_output("ss --vsock | grep %s" % port), it should work, I tested it on remote host